### ISSUE TITLE
cmd/govim: vim freezes if you delete all the text in a Go file

### ISSUE THREAD

**Issue by cespare** [govim/govim#454]
### What version of Vim/Gvim are you using?

<details><summary>Vim/Gvim version</summary><br><pre>
$ vim --version
VIM - Vi IMproved 8.1 (2018 May 18, compiled Aug 16 2019 11:12:03)
Included patches: 1-1856
Compiled by caleb@eru
Huge version with GTK2 GUI.  Features included (+) or not (-):
+acl               -farsi             -mouse_sysmouse    -tag_any_white
+arabic            +file_in_path      +mouse_urxvt       -tcl
+autocmd           +find_in_path      +mouse_xterm       +termguicolors
+autochdir         +float             +multi_byte        +terminal
-autoservername    +folding           +multi_lang        +terminfo
+balloon_eval      -footer            -mzscheme          +termresponse
+balloon_eval_term +fork()            +netbeans_intg     +textobjects
+browse            +gettext           +num64             +textprop
++builtin_terms    -hangul_input      +packages          +timers
+byte_offset       +iconv             +path_extra        +title
+channel           +insert_expand     +perl              +toolbar
+cindent           +job               +persistent_undo   +user_commands
+clientserver      +jumplist          +postscript        +vartabs
+clipboard         +keymap            +printer           +vertsplit
+cmdline_compl     +lambda            +profile           +virtualedit
+cmdline_hist      +langmap           -python            +visual
+cmdline_info      +libcall           -python3           +visualextra
+comments          +linebreak         +quickfix          +viminfo
+conceal           +lispindent        +reltime           +vreplace
+cryptv            +listcmds          +rightleft         +wildignore
+cscope            +localmap          +ruby              +wildmenu
+cursorbind        -lua               +scrollbind        +windows
+cursorshape       +menu              +signs             +writebackup
+dialog_con_gui    +mksession         +smartindent       +X11
+diff              +modify_fname      -sound             -xfontset
+digraphs          +mouse             +spell             +xim
+dnd               +mouseshape        +startuptime       +xpm
-ebcdic            +mouse_dec         +statusline        +xsmp_interact
+emacs_tags        +mouse_gpm         -sun_workshop      +xterm_clipboard
+eval              -mouse_jsbterm     +syntax            -xterm_save
+ex_extra          +mouse_netterm     +tag_binary        
+extra_search      +mouse_sgr         -tag_old_static    
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "$VIM/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/home/caleb/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/gio-unix-2.0/ -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/freetype2 -I/usr/include/libpng16   -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: gcc   -L. -L/home/caleb/.rbenv/versions/2.3.1/lib  -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,-E   -L/usr/local/lib -Wl,--as-needed -o vim   -lgtk-x11-2.0 -lgdk-x11-2.0 -lpangocairo-1.0 -latk-1.0 -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lfontconfig -lfreetype -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinfo -lnsl  -lselinux -lacl -lattr -lgpm -ldl   -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -L/usr/lib/x86_64-linux-gnu/perl/5.26/CORE -lperl -ldl -lm -lpthread -lcrypt    -Wl,-R/home/caleb/.rbenv/versions/2.3.1/lib -L/home/caleb/.rbenv/versions/2.3.1/lib -lruby-static -lpthread -ldl -lcrypt -lm  -L/home/caleb/.rbenv/versions/2.3.1/lib    

</pre></details>

### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.12.4 linux/amd64

</pre>

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GOARCH="amd64"
GOBIN="/home/caleb/bin"
GOCACHE="/home/caleb/.cache/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/caleb/p/go:/home/caleb"
GOPROXY=""
GORACE=""
GOROOT="/home/caleb/apps/go"
GOTMPDIR=""
GOTOOLDIR="/home/caleb/apps/go/pkg/tool/linux_amd64"
GCCGO="/usr/bin/gccgo"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build476200504=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

### What version/commit of `govim` are you using?
b15ac0a948cb00443cc6d80c6dd2e96a485290bd

### What is the output of `:messages` in Vim?

```
Messages maintainer: Bram Moolenaar <Bram@vim.org>
"pattern.go" 208L, 5534C
"x.go" 1L, 16C
"x.go" 0L, 0C written
```
### Do the Vim channel, `govim` or `gopls` log files show anything interesting?

Probably not relevant; I can collect them if you want.

### How did you install `govim`?

Vim 8 packages

### What did you do?

Go into a well-formed, compilable Go package. (This happens in GOPATH mode as well as a module package.)

1. Create a new file `x.go` containing only the text `package <packagename>` (matching the existing package).
2. Open `x.go` in vim and `ggdG` to delete all the contents
3. `:w` to save the file

### What did you expect to see?

Nothing should happen. Perhaps a warning in my quickfix or something since this isn't a valid Go package anymore (there's an empty .go file).

### What did you see instead?

Vim completely freezes and becomes unresponsive to input.

---

I guess a meta-issue between this and #453 is that ISTM that it shouldn't be *possible* for govim to freeze vim like this. I wonder if that indicates there's something wrong with how govim plugs into vim? I've run into tons of bad bugs with vim-go and lots of other plugins, but I rarely see those cause vim to hang.

**Comment by leitzler** [comment]
This happens due to gopls panicing (see below).

I’ve tested with latest gopls and the issue has been fixed there.

```
2019-08-21T10:22:03.875670_#1: gopls stderr: panic: runtime error: invalid memory address or nil pointer dereference
2019-08-21T10:22:03.877143_#1: gopls stderr: [signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x15343c5]
2019-08-21T10:22:03.878433_#1: gopls stderr: 
2019-08-21T10:22:03.878475_#1: gopls stderr: goroutine 132 [running]:
2019-08-21T10:22:03.878499_#1: gopls stderr: go/token.(*File).Pos(...)
2019-08-21T10:22:03.878532_#1: gopls stderr: 	/usr/local/go/src/go/token/position.go:269
2019-08-21T10:22:03.878553_#1: gopls stderr: golang.org/x/tools/internal/lsp.spanToRange(0x17bfd00, 0xc0003816e0, 0x17ca2e0, 0xc0002621c0, 0xc0000c8480, 0x51, 0x1, 0x1, 0x0, 0x1, ...)
2019-08-21T10:22:03.878575_#1: gopls stderr: 	/Users/pontus/go/pkg/mod/golang.org/x/tools@v0.0.0-20190727173135-db2fa46ec33c/internal/lsp/format.go:45 +0x215
2019-08-21T10:22:03.878598_#1: gopls stderr: golang.org/x/tools/internal/lsp.organizeImports(0x17bfd00, 0xc0003816e0, 0x17ca2e0, 0xc0002621c0, 0xc0000c8480, 0x51, 0x1, 0x1, 0x0, 0x1, ...)
2019-08-21T10:22:03.878620_#1: gopls stderr: 	/Users/pontus/go/pkg/mod/golang.org/x/tools@v0.0.0-20190727173135-db2fa46ec33c/internal/lsp/code_action.go:101 +0x9b
2019-08-21T10:22:03.878929_#1: handleEvent: Vim is in charge of /private/var/folders/xm/ls208xd174v95pgd20rn_qbh0000gn/T/tmp.nnV1MkSz/x.go; not handling 
2019-08-21T10:22:03.880927_#1: gopls stderr: golang.org/x/tools/internal/lsp.(*Server).codeAction(0xc00022e000, 0x17bfd00, 0xc0003816e0, 0xc00038c720, 0xc00038c720, 0x0, 0x0, 0x0, 0xc00022eb40)
2019-08-21T10:22:03.881433_#1: handleEvent: Vim is in charge of /private/var/folders/xm/ls208xd174v95pgd20rn_qbh0000gn/T/tmp.nnV1MkSz/x.go; not handling 
2019-08-21T10:22:03.881622_#1: gopls stderr: 	/Users/pontus/go/pkg/mod/golang.org/x/tools@v0.0.0-20190727173135-db2fa46ec33c/internal/lsp/code_action.go:50 +0x510
2019-08-21T10:22:03.881745_#1: gopls stderr: golang.org/x/tools/internal/lsp.(*Server).CodeAction(0xc00022e000, 0x17bfd00, 0xc0003816e0, 0xc00038c720, 0xc00038c720, 0x0, 0x0, 0x0, 0x0)
2019-08-21T10:22:03.881832_#1: gopls stderr: 	/Users/pontus/go/pkg/mod/golang.org/x/tools@v0.0.0-20190727173135-db2fa46ec33c/internal/lsp/server.go:204 +0x4d
2019-08-21T10:22:03.881889_#1: gopls stderr: golang.org/x/tools/internal/lsp/protocol.serverHandler.Deliver(0x17d1100, 0xc00022e000, 0x17bfd00, 0xc0003816e0, 0xc000365600, 0x0, 0xc0000e2f30)
2019-08-21T10:22:03.881949_#1: gopls stderr: 	/Users/pontus/go/pkg/mod/golang.org/x/tools@v0.0.0-20190727173135-db2fa46ec33c/internal/lsp/protocol/tsserver.go:400 +0x23ce
2019-08-21T10:22:03.882004_#1: gopls stderr: golang.org/x/tools/internal/jsonrpc2.(*Conn).Run.func1(0xc0001d8660, 0xc000365600, 0xc00021b440, 0x17bfd00, 0xc0003816e0, 0x0, 0x0, 0xc0001e7ac0)
2019-08-21T10:22:03.882045_#1: gopls stderr: 	/Users/pontus/go/pkg/mod/golang.org/x/tools@v0.0.0-20190727173135-db2fa46ec33c/internal/jsonrpc2/jsonrpc2.go:370 +0x13a
2019-08-21T10:22:03.882105_#1: gopls stderr: created by golang.org/x/tools/internal/jsonrpc2.(*Conn).Run
2019-08-21T10:22:03.882159_#1: gopls stderr: 	/Users/pontus/go/pkg/mod/golang.org/x/tools@v0.0.0-20190727173135-db2fa46ec33c/internal/jsonrpc2/jsonrpc2.go:354 +0x8d8
```

**Comment by myitcv** [comment]
Marking as on hold pending #487

**Comment by leitzler** [comment]
Tested without any issues after #487 was merged, closing.
Please upgrade govim and give it a try @cespare :)

**Comment by cespare** [comment]
Thanks!
